add target features when extracting and running doctests

QuietMisdreavus opened the pull request:
When rendering documentation, rustdoc will happily load target features into the cfg environment from the current target, but fails to do this when doing anything with doctests. This would lead to situations where, thanks to https://github.com/rust-lang/rust/pull/48759, functions tagged with `#[target_feature]` couldn't run doctests, thanks to the automatic `#[doc(cfg(target_feature = "..."))]`.

Currently, there's no way to pass codegen options to rustdoc that will affect its rustc sessions, but for now this will let you use target features that come default on the platform you're targeting.

Fixes https://github.com/rust-lang/rust/issues/49723

rust-highfive commented:
r? @steveklabnik

(rust_highfive has picked a reviewer for you, use r? to override)

QuietMisdreavus commented:
r? @GuillaumeGomez

TimNN commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/364906126) ([raw log](https://api.travis-ci.org/v3/job/364906126/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:47] configure: rust.quiet-tests     := True
---
[00:44:47] ..............................................................................i.....................
[00:44:53] .....................i..............................................................................
---
[00:45:36] i..........................................................................i........................
---
[00:46:33] .............................................i......................................................
---
[00:50:33] .............................i......................................................................
[00:50:48] ..............................................................i.....................................
[00:51:03] ...............................................i....................................................
[00:51:24] ....................................................................................................
[00:51:46] ....................................................................................................
[00:52:08] ....................................................................................................
[00:52:34] ...i...............................................................................................i
[00:53:01] .............................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:53:12] .......................
[00:53:43] ....................................................................................................
[00:54:19] .................................................................ii.................................
[00:55:05] ............................i....................................................i.ii.test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:55:12] ..............
[00:55:55] .........................................................................................iiiiiii....
---
[00:58:10] ...................i............................................................ii.iii..............
[00:58:17] ....................................................................................................
[00:58:25] .........i..............................i...........................................................
[00:58:33] ....................................................................................................
[00:58:40] .....................i..............................................................................
[00:58:48] ....................................................................................................
[00:58:58] ....................................................................................................
[00:59:08] ....................................................................................................
[00:59:19] ....................................................................................................
[00:59:33] ....................................................................................................
[00:59:42] ..............i.....................................................................................
[00:59:52] ..................i..ii.............................................................................
[01:00:02] ....................................................................................................
[01:00:12] ....................................................................................................
[01:00:22] .....................................................................................i..............
[01:00:32] ...............................i....................................................................
---
[01:01:10] ...........................i........................................................................
[01:01:12] ....................................................................i...............................
[01:01:13] ................i.......................................................
---
[01:01:28] .............i........................
---
[01:01:59] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:02:02] i.......i......................i.......
---
[01:02:41] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:02:42] ....ii...
---
[01:10:31] ....................F...............................................................................
[01:12:07] .......i............................................................................................
---
[01:12:30] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-3.9/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zmiri -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zmiri -Zunstable-options  -Lnative=/checnknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu
---
56592 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/incremental/syntax-33oa6nnkk1g08/s-f006tf4qli-10t4q2j-v8bl3ay9sjlv
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

gnzlbg commented:
Thanks, I think that for the time being this is a good stop-gap solution! 

It allows users to at least provide their own hidden environment (main, `#[target_feature]`, use static/run-time feature detection, etc.) and that is something that we will always want to provide for maximum control. 

> It'd be actually nice to pass compile options to tests... We should definitely talk about it!

If `cargo test` would set the RUSTFLAGS environment variable for doc tests that would already be great.

QuietMisdreavus commented:
Just passing `RUSTFLAGS` to doctests won't work - rustdoc handles the compilation of doctests itself, which bypasses any other command-line flags that you may want to hand to rustc. Rustdoc would need to handle the `-C` flag itself and pass those codegen options on when setting up the tests.

gnzlbg commented on src/test/rustdoc/doc-cfg-target-feature.rs:
````diff
@@ -0,0 +1,27 @@
+/// ```
+/// assert!(false);
+/// ```
+#[cfg(target_feature = "sse")]
````
If this doctest fails, is because foo was not part of the binary, so... "sse" is not set for some reason, which is very weird. It is set for x86_64 always. cc @alexcrichton

QuietMisdreavus replied:
If you run the test locally (i.e. `x.py test --stage 1 src/test/rustdoc --test-args "target-feature"`) then it will work properly, implying there's something off with the Travis setup.

alexcrichton replied:
Aha right yes, you can tag this as `// no-system-llvm`

QuietMisdreavus commented:
> Aha right yes, you can tag this as `// no-system-llvm`

Well, this makes travis skip the test at least. It still runs and passes locally on my system, too, so hopefully it works this time.

I've reverted the test to the original version (that has feature detection both on the item and in the test) and squashed it down. If travis still agrees (i.e. skips the test) then this PR is ready. I have a local branch waiting for this to add a `-C` flag to rustdoc to allow for setting codegen options, including manually setting features.

QuietMisdreavus commented:
Alright, passed travis! @GuillaumeGomez ready for review!

gnzlbg commented:
@alexcrichton what was the problem? IIRC there is some stuff with target features that we are only doing with Rust's LLVM, was that it?

alexcrichton commented:
We can only read a list of target features from our own LLVM, not system LLVMs

GuillaumeGomez commented:
Let's go then!

@bors: r+

bors commented:
:pushpin: Commit 3366032 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 3366032ab70f72763bdf45c7f1257aa20b4229ca -->
